File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.go
package main

import "fmt"

func lengthOfLongestSubstring(s string) int {
	charMap := make(map[byte]int) // < char , int >
	left := 0
	ans := 0
	for right := 0; right < len(s); right++ {
		char := s[right]
		if lastIndex, ok := charMap[char]; ok && lastIndex >= left {
			left = lastIndex + 1
		}
		charMap[char] = right

		ans = max(ans, right-left+1)
	}
	return ans
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    let left = 0
    let ans = 0
    let mp = new Map();
    for (let right = 0; right < s.length; right++) {
        if (mp.has(s[right]) && mp.get(s[right]) >= left) {
            left = mp.get(s[right])  + 1;
        }
        mp.set(s[right], right);
        ans = Math.max(ans, right - left + 1);
    }
    return ans;
};